File: programmers/연습문제/호텔 대실(level2).js
function solution(book_time) {
    book_time.forEach(([start, end], idx) => {
        let hour, minute;
        [hour, minute] = start.split(':').map(e=>+e);
        book_time[idx][0] = hour*60 + minute;
        [hour, minute] = end.split(':').map(e=>+e);
        book_time[idx][1] = hour*60 + minute;
    })
    
    book_time.sort((a, b) => a[0] - b[0])
    
    let as =[1, 2, 3, 4]
    let room = []
    book_time.forEach(([start, end]) => {
        if(!room.length) room.push(end);
        else{
            let idx = room.findIndex((v) => v+10 <= start)
            if(idx !== -1) room[idx] = end;
            else room.push(end);
        }
    })
    
    return room.length
}

File: programmers/연습문제/호텔 대실(level2).py
from heapq import heappush, heappop

def solution(book_time):
    
    for idx, [start, end] in enumerate(book_time):
        hour, minute = map(int, start.split(':'))
        book_time[idx][0] = hour*60 + minute
        hour, minute = map(int, end.split(':'))
        book_time[idx][1] = hour*60 + minute
    
    book_time.sort()

    heap = []
    
    for [start, end] in book_time:
        if(not heap): 
            heappush(heap, end)
            continue;
            
        if(heap[0]+10 <= start):
            heappop(heap)
        heappush(heap, end)
    
    return len(heap)